agents/pfeiffer_turbo_controller/Pfeiffer_Turbo_Controller_Agent.py

```diff
@@ -0,0 +1,206 @@
+        if auto_acquire:
+            self.agent.start('acq')
+
+        return True, 'Initialized Turbo Controller.'
+
+    def monitor_turbo(self, session, params=None):
+        """
+            Process to continuously monitor turbo motor temp and rotation speed and
+            send info to aggregator.
+
+            Args:
+                wait (float, optional):
+                    time to wait between measurements [seconds].
+        """
+        if params is None:
+            params = {}
+
+        wait_time = params.get('wait', 1)
+
+        self.monitor = True
+
+        while self.monitor:
+            with self.lock.acquire_timeout(1) as acquired:
+                if acquired:
+                    data = {
+                        'timestamp': time.time(),
+                        'block_name': 'turbo_output',
+                        'data': {}
+                    }
+                    
+                    try:    
+                        data['data']["Turbo_Motor_Temp"] = self.turbo.get_turbo_motor_temperature()
+                        data['data']["Rotation_Speed"] = self.turbo.get_turbo_actual_rotation_speed()
+                    
+                    except ValueError as e:
+                        self.log.error(e)
+                        
+                    # self.log.info(str(data))
+                    # print(data)
+                    self.agent.publish_to_feed('pfeiffer_turbo', data)
+
+                    # Allow this process to be queried to return current data
+                    session.data = data
+
+                else:
+                    self.log.warn("Could not acquire in monitor turbo")
+
+            time.sleep(wait_time)
+
+        return True, "Finished monitoring turbo"
+
+    def stop_monitoring(self, session, params=None):
+        """Stop monitoring the turbo output."""
+        if self.monitor:
+            self.monitor = False
+            return True, 'requested to stop taking data.'
+        else:
+            return False, 'acq is not currently running'
+
+    def turn_turbo_on(self, session, params=None):
+        """Turns the turbo on."""
+
+        with self.lock.acquire_timeout(1) as acquired:
+            if acquired:
+                self.turbo.ready_turbo()
+                time.sleep(1)
+                self.turbo.turn_turbo_motor_on()
+            else:
+                return False, "Could not acquire lock"
+
+        return True, 'Turned Turbo Motor On.'
+    
+    def turn_turbo_off(self, session, params=None):
+        """Turns the turbo off."""
+
+        with self.lock.acquire_timeout(1) as acquired:
+            if acquired:
+                self.turbo.turn_turbo_motor_off()
+                time.sleep(1)
+                self.turbo.unready_turbo()
+            else:
+                return False, "Could not acquire lock"
+
+        return True, 'Turned Turbo Motor Off.'
+                          
+    def get_turbo_error_code(self, session, params=None):
```

Comment:
Having this in a separate Task is going to mean that during normal operations, the `monitor_turbo` Process will be continuously running, an error may then occur, say the turbo overheats. The monitor process will continue to run, reporting temperature and rotation rate, but the particular error code won't be retrieved until someone goes and clicks a button to fetch the error code, or a client has to handle getting an error code, rather than the error being identified right when it happened.

I don't know the interaction with the hardware, but what happens if you check for an error code when there isn't an error? And how fast can you send queries? Would it make sense to check for errors within the monitor process and either continuously publish error codes (assuming there's one for "no error") or publish when an actual error arrises?

Reply:
I put the turbo error in the main acq process, it shouldn't take too much time